mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/massindexing/loader/MassIndexingEntityLoadingStrategy.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.mapper.pojo.massindexing.loader;
+
+import org.hibernate.search.mapper.pojo.loading.EntityLoadingTypeGroup;
+import org.hibernate.search.mapper.pojo.loading.EntityLoader;
+import org.hibernate.search.mapper.pojo.loading.EntityIdentifierScroll;
+import java.util.Set;
+
+/**
+ * A start loader for entity loading entities during mass indexing.
+ *
+ * @param <E> The resulting entity type (output)
+ * @param <O> The options for mass indexing proccess.
+ */
+public interface MassIndexingEntityLoadingStrategy<E, O extends MassIndexingOptions> {
+
+	/**
+	 * Streams the identifiers of entities to reindex.
+	 *
+	 * @param context A mass indexing context for objects to load.
+	 * @param includedTypes The expected types of loaded objects.
+	 * @return A {@link EntityIdentifierScroll}.
+	 * @throws java.lang.InterruptedException except where loading interrupted
+	 */
+	EntityIdentifierScroll createIdentifierScroll(MassIndexingThreadContext<O> context, Set<Class<? extends E>> includedTypes) throws InterruptedException;
```

Comment:
This can also move other functions here, which are also needed to compute the filtered sample for indexing: 

```java	/**
	 * Get the entity super type indexing process.
	 *
	 * @return the entity super type.
	 */
	Class<?> commonSuperType();

	/**
	 * Get the entity name of super type.
	 *
	 * @return the entity name of super type.
	 */
	String commonSuperEntityName();
```

Reply:
As a result, I get this interface form: 

```java
/**
 * A group entity types for entity loading entities during mass indexing.
 * @param <E> The resulting entity type (output)
 */
public interface MassIndexingEntityLoadingTypeGroup<E> {

	/**
	 * Get the notified group name for indexing process,
	 *
	 * @return the entity names by separated comma list.
	 */
	String notifiedGroupName();

	/**
	 * @return The names of all entity types included in this group.
	 */
	Set<String> includedEntityNames();

	/**
	 * The expected types of loaded objects
	 * @return A expected types of loaded objects.
	 */
	Set<Class<? extends E>> includedEntityTypes();

	/**
	 * @param entity An entity of any type.
	 * @return {@code true} if the given entity instance is included in this type group.
	 */
	boolean includesInstance(Object entity);
}
```
"String notifiedGroupName();" - I think it would be useful, to normalize the messages, the end artist would have access to the common name, he would not have to compute it himself with "Set <String> includedEntityNames();"; 

"Set<Class<? extends E>> includedEntityTypes();" This function is necessary, especially in mapper-orm, to build "HibernateOrmQueryLoader <> (queryFactory, includedTypesFilter)". For the enumeration of the value "includedTypesFilter". Used in "HibernateOrmJpaMassIndexingTypeLoadingStrategy".

Reply:
"boolean includesInstance(Object entity);"  Does this function only calculate that the entity is within the reported scope, or should it also check if the entity has an indexing flag. It is important to check indexing for "MassIndexingComplexHierarchyIT" test execution, some entities are not indexable although they are super class entities but not indexable. For example: 

```java
	@Indexed(enabled = false)
	public static class H2_B_D_NotIndexed extends H2_B_Indexed {

		public static final String NAME = "H2_B_D_NotIndexed";

		@GenericField
		private String dText;

		public H2_B_D_NotIndexed() {
		}

		public H2_B_D_NotIndexed(Integer id) {
			super( id );
			this.dText = "text" + id;
		}
	}
	
	@Test
	public void rootIndexed_someSubclassesIndexed_requestMassIndexingOnIndexedSubclass() {
		try ( SearchSession searchSession = createSession() ) {
			MassIndexer indexer = searchSession.massIndexer( H2_B_Indexed.class );

			backendMock.expectWorks( H2_B_Indexed.NAME, DocumentCommitStrategy.NONE, DocumentRefreshStrategy.NONE )
					.add( "4", b -> b.field( "rootText", "text4" )
					.field( "bText", "text4" ) )
					.createdThenExecuted();

			backendMock.expectIndexScaleWorks( H2_B_Indexed.NAME, searchSession.tenantIdentifier() )
					.purge()
					.mergeSegments()
					.flush()
					.refresh();

			try {
				indexer.startAndWait();
			}
			catch (InterruptedException e) {
				fail( "Unexpected InterruptedException: " + e.getMessage() );
			}
		}

		backendMock.verifyExpectationsMet();
	}	
```
If the provision of this functionality would depend solely on this method, its implementation would look like this. But the name doesn't quite fit it. 

```java
	public boolean includesInstance(Object entity, MassIndexingSessionContext sessionContext) {
		if ( !commonSuperType.javaClass().isAssignableFrom( entity.getClass() ) ) {
			return false;
		}
		return indexingContext.indexedInstance( sessionContext, commonSuperType, entity );
	}
```

Reply:
It also solves the problem by checking whether a given type of entity is indexed by filtering only those elements that can be indexed from the source. 

```java
	public void rootNotIndexed_someSubclassesIndexed_requestMassIndexingOnRoot() {
		try ( SearchSession searchSession = createSession() ) {
			MassIndexer indexer = searchSession.massIndexer( H1_Root_NotIndexed.class );

			backendMock.expectWorks( H1_B_Indexed.NAME, DocumentCommitStrategy.NONE, DocumentRefreshStrategy.NONE )
					.add( "3", b -> b.field( "rootText", "text3" )
					.field( "bText", "text3" ) )
					.createdThenExecuted();

			backendMock.expectIndexScaleWorks( H1_B_Indexed.NAME, searchSession.tenantIdentifier() )
					.purge()
					.mergeSegments()
					.flush()
					.refresh();

			try {
				indexer.startAndWait();
			}
			catch (InterruptedException e) {
				fail( "Unexpected InterruptedException: " + e.getMessage() );
			}
		}

		backendMock.verifyExpectationsMet();
	}
```

Reply:
> "String notifiedGroupName();" - I think it would be useful, to normalize the messages, the end artist would have access to the common name, he would not have to compute it himself with "Set includedEntityNames();";

I'm not a fan of this, really. But maybe you should use `toString()`?

> "Set<Class<? extends E>> includedEntityTypes();" This function is necessary, especially in mapper-orm, to build "HibernateOrmQueryLoader <> (queryFactory, includedTypesFilter)". For the enumeration of the value "includedTypesFilter". Used in "HibernateOrmJpaMassIndexingTypeLoadingStrategy".

I see. I'm not happy how we're exposing something in APIs in order to address internal needs, though...

Couldn't you just use the entity names to retrieve the corresponding `LoadingIndexedTypeContext`s, and then pass those to `HibernateOrmJpaMassIndexingTypeLoadingStrategy`?

> If the provision of this functionality would depend solely on this method, its implementation would look like this. But the name doesn't quite fit it.
> 
> ```
> 	public boolean includesInstance(Object entity, MassIndexingSessionContext sessionContext) {
> 		if ( !commonSuperType.javaClass().isAssignableFrom( entity.getClass() ) ) {
> 			return false;
> 		}
> 		return indexingContext.indexedInstance( sessionContext, commonSuperType, entity );
> 	}
> ```

A mass indexing type group only includes types which are... indexed. By definition. I think it makes sense, don't you?
So I would expect `includedEntityNames()` to only return the name of *indexed* entities. And I believe it used to. Is it no longer the case?

You only have to do this extra check because you're not really checking whether the instance's type is part of the group; you're checking if it has a common supertype with the types in this group (more or less). And then of course you have to perform extra checks.

You could change the implementation to this:

 ```
 	public boolean includesInstance(Object entity, MassIndexingSessionContext sessionContext) {
 		return includedEntityTypes().contains( sessionContext.runtimeIntrospector().detectEntityType( entity ) );
 	}
```

Reply:
Everything explains "MassIndexingComplexHierarchyIT" I rewrote this test from mapper-orm, so far it worked as this test shows. I would like to get the same functionality recreated in mapper-javabean, so the test is given in invariant functionality. It's all about making it work that way.

Reply:
I understand, but in this rewritten test, you added loaders for non-indexed classes, which is not how it works in the ORM mapper at all. In the ORM mapper, we expect a strategy to be set on each type exactly (possibly the same strategy for several types). If a type is not indexed, it simply doesn't have a loading strategy.

If your implementation is different, of course it led to workarounds. And I'm saying these workarounds are not legitimate.

In `org.hibernate.search.mapper.pojo.massindexing.impl.PojoMassIndexingIndexedTypeGroup#disjoint`, you should start with indexed types only, and aggregate them into groups. I'd expect that the resulting `MassIndexingGroups` only contains indexed types.

Reply:
The test is identical to the mapper-orm. If I choose a super type, it will index all extensions of that type if indexing is possible. Even if the super type itself is not indexable. So far it has worked in hibernate-orm. I didn't modify the test, I just moved it. I just think this action is the most advisable. I would also like to recreate it in mapper-bean. It would be bad to have a different functionality in both mappers.

Reply:
It may leave to the programmer the choice of method and give him the tools to choose his own method. 

```java
public interface MassIndexingEntityLoadingTypeGroup<E> {

	/**
	 * @return The names of all entity types included in this group.
	 */
	Set<String> includedEntityNames();

        /**
	 * @param entity An entity of any type.
	 * @return {@code true} if the given entity instance is included in this type group.
	 */
	boolean includesInstance(Object entity);

        /**
	 * @param entity An entity of any type.
	 * @return {@code true} if the given entity instance is asignable on this type super group.
	 */
	boolean assignasbleInstance(Object entity);
	
}
```

Reply:
> The test is identical to the mapper-orm.

There's at least one difference: in `org.hibernate.search.integrationtest.mapper.pojo.massindexing.MassIndexingComplexHierarchyIT#createSession`, you assign a strategy to root classes only. In the ORM mapper, we assign a strategy to each indexed type. So `H1_Root_NotIndexed` wouldn't have a loading strategy in the ORM mapper, for example.

But you're right, it apparently works anyway. **Even with my suggested patch**.

> It may leave to the programmer the choice of method and give him the tools to choose his own method.

I don't think that's a good idea. It makes the API more complex again, and we've been trying to simplify it for **weeks**. Please don't do that.

**Anyway**, this whole discussion started from my trying to guess why my change didn't work, as you claimed. But that whole discussion is irrelevant: **I just tried my patch, and it works**. I just had to change `includedEntityTypes()` to `includedTypes`. See https://github.com/wklaczynski/hibernate-search/pull/2

Please just apply the patch and let's move on? It's simpler, requires fewer SPI, and works correctly: that should be enough.

Reply:
Only that in my case of use only this way makes sense. Rather, I would rather build my own indexer externally from scratch than suggest to developers what to do with the presentation layer so that they have to replace every extension of the entity for indexing. But I can see that there may be a coincidence that someone wants to choose only from the included list, although I don't see it in my application. Maybe we will provide both options (inclused and assigned), and the one who designs his own strategy will decide for himself what is more beneficial in his case. By limiting it at this stage, we make the module useless for a large part of the solutions.

Reply:
Look, if you're going to introduce new use cases every time I point out a quirk in the code, this will go on forever.

Add tests that actually represents what you need. Without those, as far as I am concerned the use cases may not exist at all.

Reply:
All in all, everything is secured in the code so that there is no need to thoroughly check whether a given entity is indexable. It was only necessary for the total count calculation, to check if the entity should be indexed by calculating total count. Internally in the implementation, it also bypasses nonindexable entities before indexing begins. So the test will be performed without complex procedures. I think we can put two methods (included and assigned) for a programmer. Mention it in the documentation and make the "JavaBeanMapIndexingStrategy" flag for those who do not read the documentation but watch the solutions in the code. The "JavaBeanMapIndexingStrategy" itself is unlikely to be used by anyone because nobody keeps such data in the maps, but shows the diggers in the code how to solve their own problem. So I suggest adding the "includedOnly" flag, which will filter the data using included, otherwise assigned. Ultimately, the one who will build his own strategy will look at it in "JavaBeanMapIndexingStrategy" and decide what is the better solution in his case.

Reply:
I'm trying my best here but my patience has limits.

If you have a real use case for more API methods, please show me test (inside your PR) and we'll discuss it. I'm pretty sure there will be a better alternative, but we can discuss it.

If you don't have a specific, real use case in mind, stick to the simpler solution I provided.

Reply:
I do not know how to present even better than with the test "MassIndexingComplexHierarchyIT"? Which practically is a standard test in mapper-orm.

Reply:
Maybe there will be a use case where some programmer will want to limit the indexing to the entered ones, limiting the type extensions. But I don't see such a case with me. I try to make what I do useful also for me. Ultimately, I could write my own mass indexer, even mapper not worry about mapper-javabean. But I just want to help others benefit.

Reply:
> I do not know how to present even better than with the test "MassIndexingComplexHierarchyIT"? Which practically is a standard test in mapper-orm.

As I said above, this test passes with the simpler implementation and API that just checks whether the entity type is included in the group:

> **Anyway**, this whole discussion started from my trying to guess why my change didn't work, as you claimed. But that whole discussion is irrelevant: **I just tried my patch, and it works**. I just had to change `includedEntityTypes()` to `includedTypes`. See https://github.com/wklaczynski/hibernate-search/pull/2
> 
> Please just apply the patch and let's move on? It's simpler, requires fewer SPI, and works correctly: that should be enough.

Reply:
I can save according to the directions. But unfortunately for me it is a piece of useless work. And so I have to build a javabean-like mapper and my own massindexer. I know how to do it from scratch. We can approve it, but it will be a chunk of mostly useless code. I think that the operation of massindexing has already been proven and once someone came to the same conclusion by forcing the same integration test. I don't know why both solutions cannot be shared in the strategy? And how he wants to ultimately implement, he could choose a programmer dealing with building the strategy.

Reply:
But **why**? You told me your needs are expressed by `MassIndexingComplexHierarchyIT`. I give you a solution that makes `MassIndexingComplexHierarchyIT` pass. Surely your needs should be covered?

Reply:
I can finish the work according to the flaws, although for me it will be a wasted time and the resulting mass indexer will be completely useless. Eventually, I will have to build my own mapper with its own mass indexation, and in any event, it will be difficult for me to participate in the development of mapper-javabean, since I myself will not see the point of using it, because for me it becomes useless in that form. I see a lot of room for improvement in mapper-javabean, I know how to make it useful. Now I have to decide whether to use mapper-javabean at all or make my own mapper and propose to the rest of my team to use in our applications. What is there, unfortunately, is not suitable for further use. It's a pity because it's a lot of good work. I will approve what I have and leave it for thought. Maybe I can find an alternative so that I can use it, although without being able to assignes it definitely doesn't apply to me at all. I will write down a solution that assumes both possibilities. You can consider it for some time, eventually I will remove "boolean assignsInstance (Object entity)", although without it it becomes useless for me. It is a pity that so much work has been wasted. 

I really appreciate tips, they help me a lot. It made the whole thing calm from the very first prototypes. Most things can actually be done better by following your good advice. But just this possibility is important in my code for me, I suspect that indexing on the second proposed option will also turn out to be useful for someone, and maybe there will be such a case someday for me. But after asigns it is indispensable for me now. Without it, there is no way that such an indexer would be useful to me. 

I'll throw in a version that allows two alternatives and leave some time to think. The final version of the interface looks like this. This enables the execution of an indexing strategy according to two principles. Każy can choose between "boolean includesInstance (Object entity)" and "boolean assignsInstance (Object entity)". The third "boolean indexesInstance (Object entity) is helpful, but I don't insist that it is, it allows you to calculate totalCount more precisely by excluding non-indexable objects before indexing begins. 
```java
/**
 * A group entity types for entity loading entities during mass indexing.
 * @param <E> The resulting entity type (output)
 */
public interface MassIndexingEntityLoadingTypeGroup<E> {

	/**
	 * @return The names of all entity types included in this group.
	 */
	Set<String> includedEntityNames();

	/**
	 * The expected types of loaded objects
	 * @return A expected types of loaded objects.
	 */
	Set<Class<? extends E>> includedEntityTypes();

	/**
	 * @param entity An entity of any type.
	 * @return {@code true} if the given entity instance is included in this type group.
	 */
	boolean includesInstance(Object entity);

	/**
	 * @param entity An entity of any type.
	 * @return {@code true} if the given entity instance is asignable on this type super group.
	 */
	boolean assignsInstance(Object entity);

	/**
	 * @param entity An entity of any type.
	 * @return {@code true} if the given entity instance is indexed.
	 */
	boolean indexesInstance(Object entity);

}
```

Reply:
Can't you just explain why you need that? The whole code works without it! And `group.includesInstance(...)` **implies** that the instance is assignable to the common type!

Am I going mad? Are you just ignoring what I say? I don't know anymore. It's really frustrating.

Reply:
For me it is important that indexing and loading in search always return all extensions of the type similar to the queries in entity manager. And only this working search is useful in my use case. 

```java
Document1 extends Document, Document2 extends Document, ... DocumentN extends Document.


SearchResult<Document> result = session.search( scope )
		.where( (t) -> t.matchAll() )
		.fetchAll();

MassIndexer indexer = session.massIndexer( Document );
```
The search will return all documents from Document1 ... DocumentN to me, assuming every Document instance has "@Indexed (index = true)". It does not matter that the super class is @Indexed (index = true or false), or has such annotation at all.

The same mass indexing is to index all documents and it is not possible to add another implementation to the document indexer because they extended the type.

The indexer is a completely different piece of code hidden in the administrator panel. When adding another document type and extending the Document type, I have to be sure that the indexer will automatically take it into account when indexing.

As with JPA query "Select Object (o) From Document as o", it includes all documents "Document1 ... DocumentN" and similarly executes the procedure on all documents. 

Of course, I agree that there are those who would like to index only selected types explicitly specified in: 
```java
MassIndexer indexer = session.massIndexer( Document.class, Document1.class, Document2.class );
```
Then when defining Document.class, it would always index "Document1, Document2, .... DocumentN" if the policy was on "assignsInstance". So whether or not there is another document defined would not matter if it is super type. 

However, if someone would like to be able to limit it only to selected ones, and thus return only "Document, Document1, Document2", then he uses "includesInstance" in his implementation.

For me it is a valuable thing to work in "assignsInstance", but I agree that the second case can also be useful for someone.

Reply:
I think there is a misunderstanding. **What you need is exactly what is happening**. The code I suggested **does behave that way**. Really, always. We always include all relevant subtypes...

You see, before the groups are created, the selection from the user is expanded to include all indexed subtypes.

This code:

```
MassIndexer indexer = session.massIndexer( Document.class, Document1.class, Document2.class );
```

Will reindex `Document`, `Document`, `Document2`, all subtypes of `Document`, all subtypes of `Document1`, all subtypes of `Document2`. Of course it will only include the types that are actually indexed, because, well... the other types are irrelevant in this context.

The inclusion of subtypes happen very early, when we create the `SearchScope`. If you want to have a look, it happens here: `org.hibernate.search.mapper.pojo.scope.impl.PojoScopeDelegateImpl#create`. See this code:

```
		for ( PojoRawTypeIdentifier<? extends E> targetedType : targetedTypes ) {
			Optional<? extends Set<? extends PojoScopeIndexedTypeContext<?, ? extends E>>> targetedTypeManagersForType =
					indexedTypeContextProvider.allForSuperType( targetedType );
			if ( targetedTypeManagersForType.isPresent() ) {
				targetedTypeContexts.addAll( targetedTypeManagersForType.get() );
			}
			else {
				// Remember this to produce a clear error message
				nonIndexedTypes.add( targetedType );
			}
		}
```

We call `allForSuperType` to collect all indexed subtypes, and add them to the scope as well as the explicitly targeted type.

Then this set of types gets propagated and ultimately ends up in the mass indexer.

> The search will return all documents from Document1 ... DocumentN to me, assuming every Document instance has "@indexed (index = true)". It does not matter that the super class is @indexed (index = true or false), or has such annotation at all.

Please stop and think about this for a moment. If some types are not indexed, they will never appear in the search results. That's what "not indexed" means. Their identifier will never appear in the search result, even with `matchAll`, because there is no corresponding *indexed* document. So there is no reason for these types to be loaded.

In fact, the *only* reason to mark a type as `@Indexed(enabled = false)` is because you **do not want it in the search results**. If that behavior doesn't suit you, it's no big deal: just don't use `@Indexed(enabled = false)`. All evidence points to the conclusion that you don't need it.

The same goes for mass indexing. There is actually no way to index a type marked as `@Indexed(enabled = false)`. Hibernate Search will throw an exception if you try that. Even with a different mapper.

I think you got confused by the fact that we go out of our way to exclude non-indexed subtypes when loading. That is not really a feature. We do not *want* to exclude those subtypes. Generally those subtypes will not appear in search results since they are not indexed. The only reason we exclude subtypes is to **address edge cases**, where the type of an entity changed and the index wasn't updated accordingly.

Reply:
Oki. I'll correct and run the tests. I practically did not analyze "disjoint", I transferred it from the original. How it works this is what I was looking for.

Reply:
It works. It is correct. The proposed interface version is sufficient. The test "MassIndexingComplexHierarchyIT" is running correctly. 

```java
/**
 * A group entity types for entity loading entities during mass indexing.
 * @param <E> The resulting entity type (output)
 */
public interface MassIndexingEntityLoadingTypeGroup<E> {

	/**
	 * @return The names of all entity types included in this group.
	 */
	Set<String> includedEntityNames();

	/**
	 * The expected types of loaded objects
	 * @return A expected types of loaded objects.
	 */
	Set<Class<? extends E>> includedEntityTypes();

	/**
	 * @param entity An entity of any type.
	 * @return {@code true} if the given entity instance is included in this type group.
	 */
	boolean includesInstance(Object entity);

}
```

Reply:
Another bigger problem is the ignorance of the types associated with names. Rather, it is not about reading from a map. Target interfaces will retrieve entities via REST, SOAP, Spring, Ejb, etc .... At the beginning, ask a question to get the smallest possible batch of data. "includesInstance" solves the problem as the object is already loaded, load all entities later filter them using "includesInstance" will be very inefficient. you can use "includedEntityNames", but is that not enough? good to have related types. May it return some "EntityTypeReference" object?

Reply:
Maybe return the map? - Map <String, Class <? >>.

Reply:
Do you need this map for your own purposes?

If so, please add a test that uses this `includedEntityMap()` method in the JavaBean mapper.

If not, I'll remove this `includedEntityMap()` method after I merge your PR, because there are other ways to get this information in the ORM mapper.

Reply:
I know that you can "SearchMapping # indexedEntity", and determine which is the super type by various methods. In most of the cases I anticipate, the super type will be the most needed. The problem is that uploading "SearchMapping" is troublesome. Then complicate all the code to find the super type. 

And the final task is simple. Define the super type and ask the external interface (REST, SOAP, EJB, Spring, etc ...) to return a specific number of entities by the super type. 

I'm currently trying to use this indexing by coupling it to a modified JBoss picketlink. But I also have a new REST, SOAP interface, and even CORBA, where my application is working, and I need to get entities from a foreign system. But usually the super type is important in all cases. Now I hope the super type is the first to save from constantly sorting the list and extracting the super type. 

In fact, the information about the names of entity types is not important to me. I know sometimes maps are types and in those cases it will be necessary to use a name. But uploading "SearchMapping", extracting super type every time is very problematic. 

This is my working use case for pickedlink: 
```java
                session = mapping.createSessionWithOptions().
                        loading((options) -> {
                            options.documentInterceptor(WrapClassLoadingInterceptor.with(areaClassLoader));
                            options.identifierInterceptor(WrapClassLoadingInterceptor.with(areaClassLoader));
                            try {
                                SecurityHelper.with(areaClassLoader, () -> {
                                    SecurityHelper.setContextClassLoader(areaClassLoader);
                                    PortalSearchLoadingInitiator initiator = new PortalSearchLoadingInitiator(options, areaClassLoader, areaName, type, owner);

                                    List<Object> systemConfigurers = container.getSearchConfigurerMap().get(serchUnit);
                                    for (Object systemConfigurer : systemConfigurers) {
                                        if (systemConfigurer instanceof PortalSearchLoadingConfigurer) {
                                            PortalSearchLoadingConfigurer configurer = (PortalSearchLoadingConfigurer) systemConfigurer;
                                            configurer.configure(initiator);
                                        }
                                    }
                                });
                            } catch (ReflectiveOperationException ex) {
                                log.error("Can not configure loadings search session!", ex);
                            }
                        }).build();

```
Sample loader: 

```java
    @Override
    public void configure(SearchLoadingOptions options) {
        Partition partition = getPartition(options.getOwnerType(), options.getOwnerId());

        options.entityLoader(IdentityType.class, (identifiers) -> {
            IdentityManager im = createIdentityManager(partition);
            IdentityQueryBuilder qb = im.getQueryBuilder();
            IdentityQuery<IdentityType> guery = qb.createIdentityQuery(IdentityType.class);
            List<IdentityType> result = guery.where(qb.in(IdentityType.ID, identifiers)).getResultList();
            return result;
        });

        options.indexLoader(IdentityType.class, (ctx) -> {
            IdentityManager im = createIdentityManager(partition);
            IdentityQueryBuilder qb = im.getQueryBuilder();

            Class<? extends IdentityType> superType = (Class) ctx
                    .includedEntityMap().values().iterator().next();
            IdentityQuery<? extends IdentityType> guery = qb.createIdentityQuery(superType);
            guery.setHint(IdentityQueryHints.NOATTRS, true);

            guery.setLimit(ctx.batchSize());
            guery.setOffset(0);

            return EntityIdentifierScrolls.forEach(guery.getResultCount(), (offset) -> {
                if (!ctx.isActive()) {
                    throw new IndexDataException("Index data iterrupted " + ctx.processName());
                }
                guery.setOffset(offset.intValue());
                List<? extends IdentityType> result = guery.getResultList();
                return result.stream().map(IdentityType::getId).collect(Collectors.toList());
            });

        }).documentLoader((identifiers) -> {
            IdentityManager im = createIdentityManager(partition);
            IdentityQueryBuilder qb = im.getQueryBuilder();
            IdentityQuery<IdentityType> guery = qb.createIdentityQuery(IdentityType.class);
            List<IdentityType> result = guery.where(qb.in(IdentityType.ID, identifiers)).getResultList();
            return result;
        });
```
As you can see, determining the type is very important. The same applies to REST, SOAP. It is cumbersome every time to derive it from the entity name. Here again I assume that the super type is first on the list, so as not to waste time and complicate the calculation of it by sorting the elements. I just need "Set <Class <? Extends E >> includedEntityTypes ()" as in the case of mapper-orm. But combining types into a map is a very understandable solution, we have names associated with the type. 

We can go back to the proposed solution "Set <String> includedEntityNames ();" and "Set <Class <? extends E >> includedEntityTypes ();". It is also acceptable for me. But the mapper-orm itself requires the latter. I think that combining it into a map makes this solution more readable, because we know the entity names and its types.

Reply:
The second problem "entityLoader" usually does the same as "documentLoader". Maybe add a "LoadingContext" context to the document interception so that the policy programmer can follow this and call the loader from entity loading to load indexes? 

```java
public final class JavaBeanSearchLoadingContext implements PojoLoadingContext, MassIndexingContext<JavaBeanIndexingOptions>, LoadingContext {
...
	private JavaBeanSearchLoadingContext(Builder builder) {
.....
		documentInterceptors.add( ictx -> ictx.context( LoadingContext.class, this ));
	}
....
```

```java
public class MyIndexingStrategy<E> implements MassIndexingEntityLoadingStrategy<E, JavaBeanIndexingOptions> {

    @Override
    public EntityLoader<E> createLoader(MassIndexingThreadContext<JavaBeanIndexingOptions> context,
            MassIndexingEntityLoadingTypeGroup<E> loadingTypeGroup) throws InterruptedException {
            
             Class<?> superType = (Class) loadingTypeGroup
                    .includedEntityMap().values().iterator().next();

             LoadingContext loadingContext = context.context(LoadingContext.class);
             loadingContext.createLoader(superType);
            
        if (entityLoader == null) {
            throw new IllegalStateException(String.format("Entity loader for indexing for type %s can not be set.", entityTpe.getSimpleName()));
        }
        return entityLoader::load;

    }
```
It will be possible to call the entity loader in the strategy. Which will ultimately make it possible to build one loader.

Reply:
Sure, let's keep the map if it's useful for you.  But again: I'd recommend adding a test in Hibernate Search for your use case. Tests are the only way we can know how Hibernate Search is used, and keep it compatible with your needs.

Regarding your use case, though... a few comments.

First, if it's so complicated... consider not using type grouping at all?

If you disable type grouping, everything is much, much simpler, since you're guaranteed that each loader will handle a single type, and that type is always indexed.

Type grouping is just a performance optimization; it makes sense for relational databases, but in your specific case retrieving data from remote endpoints, maybe it doesn't matter?

Second, if you're going to use type grouping:

> I assume that the super type is first on the list

I'd recommend against this assumption, since we don't try to enforce this ordering and we don't have a test for that. Also, if the supertype is abstract, it just won't appear in the included types.

But really, you shouldn't need this assumption. At some point, you're assigning a loading strategy to a supertype. I.e. you call `.massIndexingLoadingStrategy( sometype, new MyMassIndexingEntityLoadingStrategy() )`  Can't you just pass the root type of your entity hierarchy to the strategy?

That's what we do in the ORM mapper:

* Each strategy is only valid for a given supertype. It knows that type because we pass it to its constructor: see `HibernateOrmEntityIdEntityLoadingStrategy#rootEntityPersister` for example.
* We assign the appropriate strategy to each entity type.
* Entity types with a common supertype will likely end up with the same strategy (or two strategy that are "equal" according to `equals`), so they will be grouped together when creating a loader.
* When we create the loader, the strategy already knows what the supertype is: it's the type targeted by `HibernateOrmEntityIdEntityLoadingStrategy#rootEntityPersister`. The strategy only needs the list of included types in case *some subtypes are excluded*. That's what `AbstractHibernateOrmLoadingStrategy#createQueryLoader` is for: if all subtypes are included, it will just ignore them and load all instances of the supertype. If some subtypes are not included, it will apply some additional filtering.

In your case, I don't think you use `@Indexed(enabled = false)`, so this probably doesn't matter. You should be able to just pass the supertype to the constructor of your loading strategy, and ignore the included types.

Reply:
> Type grouping is just a performance optimization; it makes sense for relational databases, but in your specific case retrieving data from remote endpoints, maybe it doesn't matter?

All in all, grouping is important for mapping-orm. But it also secures the indexing strategy, if the one who plans the presentation layer asks for the indexing elements of the same supertype. Then it would repeat the indexing for subtypes. All in all, a similar meaning of this solution is hibernate-orm. The default grouping is istance of, mapper-orm is different but similar. Usually, it will be rare for the developer to request a grouping other than istance of. But it can happen. Ultimately, the default type remains, and you don't have to find out this type every time. 

Maybe it would be better to set grouping on the strategy to avoid multiple indexing of elements implementing the same supertype as is the case with mapper-orm? There, too, someone foresaw such a use case and secured it in this way.

Reply:
> Sure, let's keep the map if it's useful for you. But again: I'd recommend adding a test in Hibernate Search for your use case. Tests are the only way we can know how Hibernate Search is used, and keep it compatible with your needs.

I will build a suitable test for this.

Reply:
> I'd recommend against this assumption, since we don't try to enforce this ordering and we don't have a test for that. Also, if the supertype is abstract, it just won't appear in the included types.

I will build a suitable test for this.

I assume if another supertype is shown, the first one above the type will always be the type implemented by the rest of the subtypes from the following parts of the map, but this is closely related to the selected grouping strategy. If someone has to change the clustering strategy to their own, this may not be the case. In general, a similar case is in mapper-orm, where entities can be grouped by entity name, and their type will be a map, then such order will not occur.

Reply:
> But it also secures the indexing strategy, if the one who plans the presentation layer asks for the indexing elements of the same supertype. Then it would repeat the indexing for subtypes

Yes and no... Yes it helps, but no it's not enough. The loader still needs to be careful not to blindly load all subtypes, in case some of them are annotated with `@Indexed(enabled = false)`, and thus are not included in the mass indexing type group.

But I suppose in your case it doesn't matter as long as you don't use `@Indexed(enabled = false)`.

Reply:
> * Each strategy is only valid for a given supertype. It knows that type because we pass it to its constructor: see `HibernateOrmEntityIdEntityLoadingStrategy#rootEntityPersister` for example.
> 
> * We assign the appropriate strategy to each entity type.
> 
> * Entity types with a common supertype will likely end up with the same strategy (or two strategy that are "equal" according to `equals`), so they will be grouped together when creating a loader.
> 
> * When we create the loader, the strategy already knows what the supertype is: it's the type targeted by `HibernateOrmEntityIdEntityLoadingStrategy#rootEntityPersister`. The strategy only needs the list of included types in case _some subtypes are excluded_. That's what `AbstractHibernateOrmLoadingStrategy#createQueryLoader` is for: if all subtypes are included, it will just ignore them and load all instances of the supertype. If some subtypes are not included, it will apply some additional filtering.

Initially "entityPersister" is specified, but "HibernateOrmQueryLoader # createIdentifiersQuery" filters types after "includedTypesFilter" enumerated from this grouping type map.

Reply:
> Initially "entityPersister" is specified, but "HibernateOrmQueryLoader#createIdentifiersQuery" filters types after "includedTypesFilter" enumerated from this grouping type map.

Yes, that is what I said:

> if all subtypes are included, it will just ignore them and load all instances of the supertype. If some subtypes are not included, it will apply some additional filtering.

Reply:
Usually this is how the mass indexer indexes only over type. That's enough. The programmer who designs the admin panel, will rather use a mass indexer, giving only one overtype for each entity, counting that all subtypes will be indexed. It seems to me that it would be enough to call massindexer for only one supertype. But in "SearchSession", it's different. In my use cases, too, I do not plan to specify more than one type for indexing and to call mass inexer as many times as possible over types. The grouping of types from this simply follows that you can build sessions of many types and you need to handle it somehow. Ultimately, in most cases, "includedEntityMap()" will always be one type. But rather, you should not be able to index only the highest syper type, if the programmer orders to index only the selected part by specifying only one super type.

```java

class User extends IdentityType 
class Group extends IdentityType
class Role extends IdentityType

//Indexes the User entities and their subtypes, if any. 
//"includedEntityMap" only contains "User" entity type. 
MassIndexer indexer = session.massIndexer( User.class );

//Indexes the Group entities and their subtypes, if any. 
//"includedEntityMap" only contains "Group" entity type. 
MassIndexer indexer = session.massIndexer( Group.class );

//Indexes the Role entities and their subtypes, if any. 
//"includedEntityMap" only contains "Role" entity type. 
MassIndexer indexer = session.massIndexer( Role.class );

//Indexes the User, Group, Role entities and their subtypes, if any. 
//Each case will be in a different group. 
//"includedEntityMap" contains only one of ("Role", "Group","User") item depending on the group. 
MassIndexer indexer = session.massIndexer( User.class, Group.class,  Role.class);

//Indexes the User entities and their subtypes, if any. 
//"includedEntityMap" only contains "IdentityType" entity type. 
//Indexes all subtypes, even the above cases. 
MassIndexer indexer = session.massIndexer( IdentityType.class );

//Indexes the IdentityType entities and their subtypes, if any. 
//"includedEntityMap" is included in the first "IdentityType", and in the next items (User, Group, Role). 
//Indexes all subtypes, even the above cases. It is not necessary to enumerate overtypes. Unless the indexing strategy //developer has anticipated something else and uses subtypes in the query. 
MassIndexer indexer = session.massIndexer( User.class, Group.class,  Role.class, IdentityType.class);

//All these cases can be obtained with one indexing strategy for "IdentityType".
//It is not necessary to build separate strategies but for each possible type of entity.
//There is even no need to define a strategy for the next subtype. 

    public EntityIdentifierScroll createIdentifierScroll(MassIndexingThreadContext<JavaBeanIndexingOptions> context,
            MassIndexingEntityLoadingTypeGroup<E> loadingTypeGroup) throws InterruptedException {

            IdentityManager im = createIdentityManager(partition);
            IdentityQueryBuilder qb = im.getQueryBuilder();

            //Supertype to get entities only for a given "session.massIndexer (...)" supertype
            //for a given grouping indexing supertype request. 
            Class<? extends IdentityType> superType = (Class) loadingTypeGroup
                    .includedEntityMap().values().iterator().next();
                  
            //return only for supertytpe from  "session.massIndexer (...)"  
            IdentityQuery<? extends IdentityType> guery = qb.createIdentityQuery(superType);
            guery.setHint(IdentityQueryHints.NOATTRS, true);

            guery.setLimit(ctx.batchSize());
            guery.setOffset(0);

            return EntityIdentifierScrolls.forEach(guery.getResultCount(), (offset) -> {
                if (!ctx.isActive()) {
                    throw new IndexDataException("Index data iterrupted " + ctx.processName());
                }
                guery.setOffset(offset.intValue());
                List<? extends IdentityType> result = guery.getResultList();
                return result.stream().map(IdentityType::getId).collect(Collectors.toList());
            });
    }

```

Reply:
> Yes, that is what I said:
> 
> > if all subtypes are included, it will just ignore them and load all instances of the supertype. If some subtypes are not included, it will apply some additional filtering.

Maybe in mapper-javabean let the one who designs the indexing strategy solve it? He has a type map of entities and their names in the group, he can use this and design a use case according to his own needs. It is difficult to predict all use cases.

Reply:
Usually this is how the mass indexer indexes only over type. That's enough. The programmer who designs the admin panel, will rather use a mass indexer, giving only one overtype for each entity, counting that all subtypes will be indexed.

> It seems to me that it would be enough to call massindexer for only one supertype.

Yes.

> Ultimately, in most cases, "includedEntityMap()" will always be one type.

That's only true if the given type has no subtype...

If grouping is enabled and the given type has indexed subtypes, `includedEntityMap` will include every indexed subtype. If the given type is abstract, it won't even be included in that map.

You can check that yourself. Just run `org.hibernate.search.integrationtest.mapper.pojo.massindexing.MassIndexingComplexHierarchyIT#rootIndexed_someSubclassesIndexed_requestMassIndexingOnRoot` and see what's the value of `loadingTypeGroup` when `org.hibernate.search.mapper.javabean.massindexing.impl.JavaBeanMapIndexingStrategy#createIdentifierScroll` is called. 

In that test, you're lucky: the root type is not abstract, so it's in the list of included types. But if it was abstract, it wouldn't be in the list. And again, order is not guaranteed anywhere in the API. The fact that the root type appears first is an implementation detail.

No, really: type grouping and included types are not what you think they are. The loading strategy is expected to have some knowledge of the type hierarchy it handles; it cannot blindly rely on `includedEntityMap`.

> But rather, you should not be able to index only the highest syper type, if the programmer orders to index only the selected part by specifying only one super type.

If the programmer specifies a supertype, we will index all (indexed) subtypes as well. That's how it works right now, and I see no reason to change that.

What's your point... ?

Reply:
> Maybe in mapper-javabean let the one who designs the indexing strategy solve it? He has a type map of entities and their names in the group, he can use this and design a use case according to his own needs. It is difficult to predict all use cases.

Sure, the loader will have to do the filtering itself. We'll keep that map, I'm not arguing about that. Just explaining that in your strategy implementation the root type should be known from the strategy regardless of the included types.

Anyway... whatever. Let's leave it at that, we've both lost enough time on this :)

Please address the other comments and the [Sonar warnings](https://sonarcloud.io/project/issues?id=org.hibernate.search%3Ahibernate-search-parent&pullRequest=2503&resolved=false&rules=java%3AS3740&types=CODE_SMELL) and then I'll review again, and hopefully merged.

Reply:
> But I suppose in your case it doesn't matter as long as you don't use `@Indexed(enabled = false)`.

This is also checked internally during indexing. Simply here, checking "includesInstance" is only useful for specifying totalCount, so in some cases not to include non-indexable entities for total count calculation and unnecessary loading of ignored entities while indexing. Therefore, you can restrict the id's of unloaded entities when enumerating the id's with "includesInstance". But rather asking such a detailed query to different interfaces (REST, SOAP, etc ...) would be baseline and hard to do, with my example I didn't use "includesInstance", but the function is most useful. Additionally, this is what the "MassIndexingComplexHierarchyIT#rootIndexed_someSubclassesIndexed_requestMassIndexingOnIndexedSubclass" test is examining.

Reply:
> > Ultimately, in most cases, "includedEntityMap()" will always be one type.
> 
> That's only true if the given type has no subtype...

I checked it, it is always the closest to the type of selected entities in "session.massIndexer (...);". It is set as the first type in "includedEntityMap". 
It even examines this test "MassIndexingComplexHierarchyIT#rootIndexed_someSubclassesIndexed_requestMassIndexingOnIndexedSubclass" rewritten from mapper-orm. 

So I assume that globally this is higher than the entity type. And the group supertype enumerated from the group is listed first in "includedEntityMap", but may be a subtype of the highest type supported by the given indexing strategy.

But it would really be nice to do tests for "includedEntityMap", depending on the query for "session.massIndexer (....)".

Reply:
I will write a test to check the correctness "includedEntityMap", the order of the types of the super types first and what types it returns depending on "massIndexer (...)".